camel-core/src/main/java/org/apache/camel/processor/RestBindingAdvice.java

```diff
@@ -167,11 +168,8 @@ private void unmarshal(Exchange exchange, Map<String, Object> state) throws Exce
             isJson = consumes != null && consumes.toLowerCase(Locale.ENGLISH).contains("json");
         }
 
-        // set the INPUT_TYPE to indicate body type
-        if (isJson) {
-            exchange.setProperty(Exchange.INPUT_TYPE, new DataType("json"));
-        } else if (isXml) {
-            exchange.setProperty(Exchange.INPUT_TYPE, new DataType("xml"));
+        if (exchange.getIn() instanceof DataTypeAware && (isJson || isXml)) {
+            ((DataTypeAware)exchange.getIn()).setDataType(new DataType(isJson ? "json" : "xml"));
```

Comment:
json and xml literals are used in several places. You can use a constant, it will avoid potential typo and duplicated code